xdsl/dialects/func.py

```diff
@@ -47,6 +56,97 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            lambda: parser.parse_optional_argument()
+            or parser.parse_optional_type()
+            or parser.raise_error("Expected argument or type"),
+        )
+
+        # Check consistency (They should be either all named or none)
+        if isa(args, list[parser.Argument]):
+            entry_args = args
+            input_types = [a.type for a in args]
+            assert isa(input_types, list[Attribute])
+        elif isa(args, list[Attribute]):
+            entry_args = None
+            input_types = args
+        else:
+            parser.raise_error(
+                "Expected all arguments to be named or all arguments to be unnamed."
+            )
+
+        # Parse return type
+        if parser.parse_optional_punctuation("->"):
+            return_types = parser.parse_optional_type()
+            if return_types:
+                return_types = [return_types]
+            else:
+                return_types = parser.parse_comma_separated_list(
+                    parser.Delimiter.PAREN, parser.parse_type
+                )
+        else:
+            return_types = []
+
+        attr_dict = parser.parse_optional_attr_dict_with_keyword(
+            ("sym_name", "function_type", "sym_visibility")
+        )
+
+        # Parse body
+        region = parser.parse_optional_region(entry_args) or Region()
+        func = FuncOp.from_region(name, input_types, return_types, region, visibility)
+        if attr_dict is not None:
+            func.attributes |= attr_dict.data
+        return func
+
+    def print(self, printer: Printer):
+        if self.sym_visibility:
+            visibility = self.sym_visibility.data
+            printer.print(f" {visibility}")
+
+        printer.print(f" @{self.sym_name.data}")
+        if len(self.body.blocks) > 0:
+            printer.print("(")
+            printer.print_list(self.body.blocks[0].args, printer.print_block_argument)
+            printer.print(") ")
+            if len(self.function_type.outputs) > 0:
```

Comment:
I recently discovered that PEP8 actually recommends this spelling:
```suggestion
            if self.function_type.outputs:
```
I like it as it lets the data structure optimise the is empty check and not compute the length